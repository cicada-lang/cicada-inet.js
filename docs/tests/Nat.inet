defcons zero 0 1 end

defcons add1 1 1 end

defelim add 2 1 end

defru zero add end

defru add1 add add add1 end

defnet two
  zero add1
  zero add1
  add
end

defnet four
  two two add
end