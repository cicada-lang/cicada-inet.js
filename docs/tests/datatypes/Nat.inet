type Nat 0 end
node zero -- value!: Nat end
node add1 prev: Nat -- value!: Nat end

node add target!: Nat; addend: Nat -- return: Nat end

rule zero add
  [add]-addend
  return-[add]
end

rule add1 add
  [add]-addend [add1]-prev add add1
  return-[add]
end

show zero zero add end
run zero zero add end

claim one Nat end
define one zero add1 end

claim two Nat end
define two one one add end

claim four Nat end
define four two two add end

show two end
run two end

show four end
run four end

show zero one add end
run zero one add end

claim addadd Nat Nat Nat -- Nat end
define addadd add add end

run one one one addadd end

// To define `mul`, we first need `nat_erase` and `nat_dup`.

node nat_erase target!: Nat -- end

rule zero nat_erase end
rule add1 nat_erase [add1]-prev nat_erase end

show two nat_erase end
run two nat_erase end

node nat_dup target!: Nat -- second: Nat; first: Nat end

rule zero nat_dup
  zero first-[nat_dup]
  zero second-[nat_dup]
end

rule add1 nat_dup
  [add1]-prev nat_dup $first $second
  first add1 first-[nat_dup]
  second add1 second-[nat_dup]
end

show two nat_dup end
run two nat_dup end

node mul target!: Nat; mulend: Nat -- return: Nat end

rule zero mul
  [mul]-mulend nat_erase
  zero return-[mul]
end

rule add1 mul
  [mul]-mulend nat_dup $first $second
  [add1]-prev first mul second add
  return-[mul]
end

show two two mul end
run two two mul end

claim three Nat end
define three two one add end

show three three mul end
run three three mul end

// To define `max`, we need `max_aux`.

node max_aux left: Nat; right!: Nat -- return: Nat end

node max left!: Nat; right: Nat -- return: Nat end

rule zero max
  [max]-right return-[max]
end

rule add1 max
  [max]-right [add1]-prev max_aux
  return-[max]
end

rule zero max_aux
  [max_aux]-left add1
  return-[max_aux]
end

rule add1 max_aux
  [add1]-prev [max_aux]-left max
  add1 return-[max_aux]
end


run zero two max end
run one two max end
run three two max end