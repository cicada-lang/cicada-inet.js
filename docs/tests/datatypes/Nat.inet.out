(zero₁)-value!target-(add₀)
(zero₀)-value addend-(add₀)
(add₀)-return

(zero₂)-value

(add1₁)-value!target-(add₂)
(add1₀)-value addend-(add₂)
(zero₅)-value prev-(add1₁)
(zero₄)-value prev-(add1₀)
(add₂)-return

(add1₂)-value prev-(add1₄)
(zero₆)-value prev-(add1₂)
(add1₄)-value

(add1₈)-value!target-(add₆)
(add1₆)-value!target-(add₅)
(add₅)-return addend-(add₇)
(add₆)-return target-(add₇)
(add1₇)-value addend-(add₆)
(zero₁₁)-value prev-(add1₈)
(zero₁₀)-value prev-(add1₇)
(add1₅)-value addend-(add₅)
(zero₉)-value prev-(add1₆)
(zero₈)-value prev-(add1₅)
(add₇)-return

(add1₉)-value prev-(add1₁₆)
(add1₁₆)-value prev-(add1₁₅)
(add1₁₅)-value prev-(add1₁₄)
(zero₁₂)-value prev-(add1₉)
(add1₁₄)-value

(add1₁₇)-value!target-(add₁₅)
(zero₁₆)-value addend-(add₁₅)
(zero₁₇)-value prev-(add1₁₇)
(add₁₅)-return

(zero₁₈)-value prev-(add1₁₉)
(add1₁₉)-value

(add1₂₀)-value prev-(add1₂₅)
(add1₂₅)-value prev-(add1₂₄)
(zero₂₀)-value prev-(add1₂₀)
(add1₂₄)-value

(add1₂₇)-value!target-(add₂₃)
(add₂₃)-return target-(nat_erase₀)
(add1₂₆)-value addend-(add₂₃)
(zero₂₄)-value prev-(add1₂₇)
(zero₂₃)-value prev-(add1₂₆)



(add1₃₂)-value!target-(add₂₆)
(add₂₆)-return target-(nat_dup₀)
(add1₃₁)-value addend-(add₂₆)
(zero₂₈)-value prev-(add1₃₂)
(zero₂₇)-value prev-(add1₃₁)
(nat_dup₀)-second
(nat_dup₀)-first

(zero₃₂)-value prev-(add1₃₉)
(zero₃₁)-value prev-(add1₃₈)
(add1₃₉)-value prev-(add1₃₇)
(add1₃₈)-value prev-(add1₃₆)
(add1₃₆)-value
(add1₃₇)-value

(add1₄₃)-value!target-(add₃₀)
(add1₄₁)-value!target-(add₂₉)
(add₂₉)-return mulend-(mul₀)
(add₃₀)-return target-(mul₀)
(add1₄₂)-value addend-(add₃₀)
(zero₃₆)-value prev-(add1₄₃)
(zero₃₅)-value prev-(add1₄₂)
(add1₄₀)-value addend-(add₂₉)
(zero₃₄)-value prev-(add1₄₁)
(zero₃₃)-value prev-(add1₄₀)
(mul₀)-return

(zero₄₅)-value prev-(add1₆₁)
(add1₅₅)-value prev-(add1₅₈)
(add1₆₁)-value prev-(add1₅₅)
(add1₅₈)-value prev-(add1₅₂)
(add1₅₂)-value

(add1₆₇)-value!target-(add₄₆)
(add1₆₆)-value!target-(add₄₅)
(add1₆₄)-value!target-(add₄₄)
(add1₆₃)-value!target-(add₄₃)
(add₄₄)-return mulend-(mul₆)
(add₄₆)-return target-(mul₆)
(add₄₅)-return addend-(add₄₆)
(zero₅₅)-value prev-(add1₆₇)
(add1₆₅)-value addend-(add₄₅)
(zero₅₄)-value prev-(add1₆₆)
(zero₅₃)-value prev-(add1₆₅)
(add₄₃)-return addend-(add₄₄)
(zero₅₂)-value prev-(add1₆₄)
(add1₆₂)-value addend-(add₄₃)
(zero₅₁)-value prev-(add1₆₃)
(zero₅₀)-value prev-(add1₆₂)
(mul₆)-return

(zero₆₈)-value prev-(add1₁₀₄)
(add1₉₅)-value prev-(add1₁₀₁)
(add1₈₂)-value prev-(add1₉₈)
(add1₁₀₄)-value prev-(add1₉₅)
(add1₁₀₁)-value prev-(add1₉₂)
(add1₉₈)-value prev-(add1₈₉)
(add1₉₂)-value prev-(add1₈₅)
(add1₈₉)-value prev-(add1₇₉)
(add1₈₅)-value prev-(add1₈₂)
(add1₇₉)-value

(add1₁₀₅)-value prev-(add1₁₀₈)
(zero₇₆)-value prev-(add1₁₀₅)
(add1₁₀₈)-value

(add1₁₁₄)-value prev-(add1₁₁₃)
(zero₇₉)-value prev-(add1₁₁₄)
(add1₁₁₃)-value

(add1₁₁₅)-value prev-(add1₁₂₄)
(add1₁₂₄)-value prev-(add1₁₂₂)
(zero₈₁)-value prev-(add1₁₁₅)
(add1₁₂₂)-value

